<a href="https://colab.research.google.com/github/ithapa1/data-visualization/blob/main/Exam2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exam 2**


Indra Thapa

CPSMA 4313

In [128]:
import re
from scipy import stats 
import matplotlib.pyplot as plt
import numpy as np
from bs4 import BeautifulSoup
import pandas as pa
df = pa.read_csv('https://raw.githubusercontent.com/nurfnick/Data_Viz/main/Activity_Dataset_V1.csv')
df.head(10)

Unnamed: 0 activity_day     workout_type  distance  time  calories  \
0           0   2022-01-01        Freestyle      9.30    77       123   
1           1   2022-01-01        Freestyle      3.44    96        55   
2           2   2022-01-01   Indoor Cycling      6.34    85        33   
3           3   2022-01-01          Walking      7.91    42        82   
4           4   2022-01-01       Open Water      8.99    36       131   
5           5   2022-01-01  Outdoor Cycling      5.47   117       144   
6           6   2022-01-01   Indoor Cycling      6.28    66       378   
7           7   2022-01-01  Outdoor Running      2.94    26       430   
8           8   2022-01-01        Trail Run      2.87    26       465   
9           9   2022-01-01    Pool Swimming      3.91    23        99   

   total_steps  avg_speed  avg_cadence  max_cadence  ... max_pace min_pace  \
0          NaN      18.88       168.54       138.30  ...      NaN      NaN   
1          NaN      29.65       125.92       292.81  ...      NaN      NaN   
2          NaN      17.85        81.93       323.69  ...      NaN      NaN   
3       1571.0      22.10        29.63       180.16  ...    28:58    07:58   
4          NaN      25.83        64.55       342.89  ...      NaN      NaN   
5          NaN       2.20        43.47       291.67  ...      NaN      NaN   
6          NaN      10.65        23.83       253.84  ...      NaN      NaN   
7       7618.0      10.59       131.93       334.72  ...    27:44    06:15   
8       2790.0      32.92        21.42       341.17  ...    25:50    06:20   
9          NaN      33.15       169.30       206.97  ...      NaN      NaN   

  avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0          112.5           122.0             103          19          28   
1          111.0           122.0             100          42          28   
2           95.0            90.0             100           1          32   
3           83.0            85.0              81           3          22   
4          138.0           166.0             110           7           0   
5          142.0           177.0             107          44          45   
6           85.0            90.0              80          34          32   
7          145.0           190.0             100          23           0   
8          142.0           188.0              96           2           0   
9          132.5           175.0              90          22          20   

   anaerobic(%)  intensive(%)  light(%)  
0             2             7        50  
1             2            29        88  
2             0            22        43  
3             0            24        65  
4             5            21        88  
5            15            57        39  
6             0            44        23  
7            71            18         3  
8            80            46        13  
9            15            28        20  

[10 rows x 21 columns]

In [129]:
df.dtypes

Unnamed: 0          int64
activity_day       object
workout_type       object
distance          float64
time                int64
calories            int64
total_steps       float64
avg_speed         float64
avg_cadence       float64
max_cadence       float64
avg_pace           object
max_pace           object
min_pace           object
avg_heart_rate    float64
max_heart_rate    float64
min_heart_rate      int64
vo2_max(%)          int64
aerobic(%)          int64
anaerobic(%)        int64
intensive(%)        int64
light(%)            int64
dtype: object

Now, I am going to delete the unnamed column from the given tables. To remove the unnamed column I am going to use the drop function.

In [130]:
df1=df.drop('Unnamed: 0', axis =1)
df1

activity_day    workout_type  distance  time  calories  total_steps  \
0     2022-01-01       Freestyle      9.30    77       123          NaN   
1     2022-01-01       Freestyle      3.44    96        55          NaN   
2     2022-01-01  Indoor Cycling      6.34    85        33          NaN   
3     2022-01-01         Walking      7.91    42        82       1571.0   
4     2022-01-01      Open Water      8.99    36       131          NaN   
..           ...             ...       ...   ...       ...          ...   
995   2022-02-14   Pool Swimming      6.92   101       313          NaN   
996   2022-02-14       Freestyle      9.48    61       219          NaN   
997   2022-02-14       Freestyle      3.30    40       169          NaN   
998   2022-02-14  Indoor Cycling      4.40    47        52          NaN   
999   2022-02-14  Indoor Cycling      1.57   116       106          NaN   

     avg_speed  avg_cadence  max_cadence avg_pace max_pace min_pace  \
0        18.88       168.54       138.30      NaN      NaN      NaN   
1        29.65       125.92       292.81      NaN      NaN      NaN   
2        17.85        81.93       323.69      NaN      NaN      NaN   
3        22.10        29.63       180.16    07:58    28:58    07:58   
4        25.83        64.55       342.89      NaN      NaN      NaN   
..         ...          ...          ...      ...      ...      ...   
995      28.66       173.30       234.73      NaN      NaN      NaN   
996       9.23       114.34       109.32      NaN      NaN      NaN   
997       7.37       133.97       122.96      NaN      NaN      NaN   
998      10.75       147.65       116.51      NaN      NaN      NaN   
999      25.37       143.41       182.80      NaN      NaN      NaN   

     avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0             112.5           122.0             103          19          28   
1             111.0           122.0             100          42          28   
2              95.0            90.0             100           1          32   
3              83.0            85.0              81           3          22   
4             138.0           166.0             110           7           0   
..              ...             ...             ...         ...         ...   
995           133.5           175.0              92          22          20   
996           103.0           122.0              84          18          28   
997           114.5           122.0             107          18          28   
998            96.0            90.0             102          45          32   
999            96.5            90.0             103          14          32   

     anaerobic(%)  intensive(%)  light(%)  
0               2             7        50  
1               2            29        88  
2               0            22        43  
3               0            24        65  
4               5            21        88  
..            ...           ...       ...  
995            15            21        61  
996             2            17        86  
997             2            23        98  
998             0            13        57  
999             0            24        83  

[1000 rows x 20 columns]

Now, this one is the filtered activities after deleting the % and spaces between them.

In [131]:
cols = df1.columns.map(lambda s: re.sub(r"\(%\)","",s))
cols

Index(['activity_day', 'workout_type', 'distance', 'time', 'calories',
       'total_steps', 'avg_speed', 'avg_cadence', 'max_cadence', 'avg_pace',
       'max_pace', 'min_pace', 'avg_heart_rate', 'max_heart_rate',
       'min_heart_rate', 'vo2_max', 'aerobic', 'anaerobic', 'intensive',
       'light'],
      dtype='object')

Here, I am going to get the datetime of the given activities which in the orginal format.

Let's see how it looks into the old table after deleting the %, spaces and unnamed.

In [132]:
df1.head()

activity_day    workout_type  distance  time  calories  total_steps  \
0   2022-01-01       Freestyle      9.30    77       123          NaN   
1   2022-01-01       Freestyle      3.44    96        55          NaN   
2   2022-01-01  Indoor Cycling      6.34    85        33          NaN   
3   2022-01-01         Walking      7.91    42        82       1571.0   
4   2022-01-01      Open Water      8.99    36       131          NaN   

   avg_speed  avg_cadence  max_cadence avg_pace max_pace min_pace  \
0      18.88       168.54       138.30      NaN      NaN      NaN   
1      29.65       125.92       292.81      NaN      NaN      NaN   
2      17.85        81.93       323.69      NaN      NaN      NaN   
3      22.10        29.63       180.16    07:58    28:58    07:58   
4      25.83        64.55       342.89      NaN      NaN      NaN   

   avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0           112.5           122.0             103          19          28   
1           111.0           122.0             100          42          28   
2            95.0            90.0             100           1          32   
3            83.0            85.0              81           3          22   
4           138.0           166.0             110           7           0   

   anaerobic(%)  intensive(%)  light(%)  
0             2             7        50  
1             2            29        88  
2             0            22        43  
3             0            24        65  
4             5            21        88

In [133]:
df1.dtypes

activity_day       object
workout_type       object
distance          float64
time                int64
calories            int64
total_steps       float64
avg_speed         float64
avg_cadence       float64
max_cadence       float64
avg_pace           object
max_pace           object
min_pace           object
avg_heart_rate    float64
max_heart_rate    float64
min_heart_rate      int64
vo2_max(%)          int64
aerobic(%)          int64
anaerobic(%)        int64
intensive(%)        int64
light(%)            int64
dtype: object

Now, I am going to get the datetime format in a original form.

In [134]:
pa.to_datetime(df1.activity_day, format='%Y-%m-%d')


0     2022-01-01
1     2022-01-01
2     2022-01-01
3     2022-01-01
4     2022-01-01
         ...    
995   2022-02-14
996   2022-02-14
997   2022-02-14
998   2022-02-14
999   2022-02-14
Name: activity_day, Length: 1000, dtype: datetime64[ns]

Now, to see a changes in an activity_day, I am going to pull the data types again in df1 which is in new data format. In the following table we can see the activity_day is in datetime format.

In [135]:
df1.dtypes

activity_day       object
workout_type       object
distance          float64
time                int64
calories            int64
total_steps       float64
avg_speed         float64
avg_cadence       float64
max_cadence       float64
avg_pace           object
max_pace           object
min_pace           object
avg_heart_rate    float64
max_heart_rate    float64
min_heart_rate      int64
vo2_max(%)          int64
aerobic(%)          int64
anaerobic(%)        int64
intensive(%)        int64
light(%)            int64
dtype: object

In [136]:
df1.activity_day = pa.to_datetime(df1.activity_day, format='%Y-%m-%d')

Now, I am going to figure out the total numbers of NaNs in the dataset.

In [137]:
df1.isna().sum().sum()

1784

Here, I am going to fill out all the 1784 NaNs in the form of '0'

In [138]:
df1.total_steps.fillna(0)

0         0.0
1         0.0
2         0.0
3      1571.0
4         0.0
        ...  
995       0.0
996       0.0
997       0.0
998       0.0
999       0.0
Name: total_steps, Length: 1000, dtype: float64

In [139]:
df1.total_steps = df1.total_steps.fillna(0)# to apply in the original table we gotta put in the given format to see changes.

Now, we can see the 0's in all the NaNs place.

In [140]:
df1.avg_pace = df1.avg_pace.fillna(0)

here, all the NaNs value in the avg_pace has been replaced by '0' as question asked and the given formula is to change the NaNs into the '0' form.

In [141]:
df1.head()

activity_day    workout_type  distance  time  calories  total_steps  \
0   2022-01-01       Freestyle      9.30    77       123          0.0   
1   2022-01-01       Freestyle      3.44    96        55          0.0   
2   2022-01-01  Indoor Cycling      6.34    85        33          0.0   
3   2022-01-01         Walking      7.91    42        82       1571.0   
4   2022-01-01      Open Water      8.99    36       131          0.0   

   avg_speed  avg_cadence  max_cadence avg_pace max_pace min_pace  \
0      18.88       168.54       138.30        0      NaN      NaN   
1      29.65       125.92       292.81        0      NaN      NaN   
2      17.85        81.93       323.69        0      NaN      NaN   
3      22.10        29.63       180.16    07:58    28:58    07:58   
4      25.83        64.55       342.89        0      NaN      NaN   

   avg_heart_rate  max_heart_rate  min_heart_rate  vo2_max(%)  aerobic(%)  \
0           112.5           122.0             103          19          28   
1           111.0           122.0             100          42          28   
2            95.0            90.0             100           1          32   
3            83.0            85.0              81           3          22   
4           138.0           166.0             110           7           0   

   anaerobic(%)  intensive(%)  light(%)  
0             2             7        50  
1             2            29        88  
2             0            22        43  
3             0            24        65  
4             5            21        88

Now, we can see the total_steps and avg_pace's NaNs value has been replaced by the '0'.

In [142]:
df1.info

<bound method DataFrame.info of     activity_day    workout_type  distance  time  calories  total_steps  \
0     2022-01-01       Freestyle      9.30    77       123          0.0   
1     2022-01-01       Freestyle      3.44    96        55          0.0   
2     2022-01-01  Indoor Cycling      6.34    85        33          0.0   
3     2022-01-01         Walking      7.91    42        82       1571.0   
4     2022-01-01      Open Water      8.99    36       131          0.0   
..           ...             ...       ...   ...       ...          ...   
995   2022-02-14   Pool Swimming      6.92   101       313          0.0   
996   2022-02-14       Freestyle      9.48    61       219          0.0   
997   2022-02-14       Freestyle      3.30    40       169          0.0   
998   2022-02-14  Indoor Cycling      4.40    47        52          0.0   
999   2022-02-14  Indoor Cycling      1.57   116       106          0.0   

     avg_speed  avg_cadence  max_cadence avg_pace max_pace min_pace

In [150]:
df1.groupby('workout_type').agg(['mean','median','count','std'])

distance                              time               \
                     mean median count       std       mean median count   
workout_type                                                               
Cricket          5.283763  5.150    93  2.847169  68.430108   67.0    93   
Freestyle        5.044792  4.505    96  2.799025  70.822917   70.0    96   
Indoor Cycling   5.259000  5.050    80  2.413489  70.025000   76.0    80   
Open Water       5.180659  4.660    91  2.767597  69.241758   69.0    91   
Outdoor Cycling  5.444000  5.570    85  2.497849  73.882353   80.0    85   
Outdoor Running  5.048025  4.500    81  2.407562  71.790123   72.0    81   
Pool Swimming    5.368085  4.840    94  2.900729  75.978723   76.5    94   
Trail Run        5.148000  5.145    90  2.519667  71.722222   72.5    90   
Treadmill        4.999898  4.730    98  2.703583  72.163265   68.5    98   
Trekking         5.278617  5.250    94  2.614193  70.308511   67.0    94   
Walking          5.612653  5.675    98  2.613102  68.642857   69.0    98   

                              calories         ... anaerobic(%)       \
                       std        mean median  ...        count  std   
workout_type                                   ...                     
Cricket          26.748181  307.548387  330.0  ...           93  0.0   
Freestyle        29.001268  278.552083  294.0  ...           96  0.0   
Indoor Cycling   29.268714  280.450000  306.0  ...           80  0.0   
Open Water       30.018639  296.747253  328.0  ...           91  0.0   
Outdoor Cycling  30.630055  299.129412  301.0  ...           85  0.0   
Outdoor Running  31.535582  301.469136  349.0  ...           81  0.0   
Pool Swimming    28.728233  283.414894  300.0  ...           94  0.0   
Trail Run        27.490970  267.966667  264.0  ...           90  0.0   
Treadmill        30.361972  278.142857  269.5  ...           98  0.0   
Trekking         30.234221  283.127660  280.5  ...           94  0.0   
Walking          29.416052  276.040816  270.5  ...           98  0.0   

                intensive(%)                           light(%)               \
                        mean median count        std       mean median count   
workout_type                                                                   
Cricket            36.924731   38.0    93  17.329440  58.849462   67.0    93   
Freestyle          37.635417   38.5    96  17.391540  49.562500   48.0    96   
Indoor Cycling     39.600000   40.0    80  18.815291  51.100000   52.0    80   
Open Water         38.153846   41.0    91  17.512486  50.252747   53.0    91   
Outdoor Cycling    33.470588   35.0    85  19.136470  49.247059   51.0    85   
Outdoor Running    32.888889   30.0    81  18.407200  55.160494   57.0    81   
Pool Swimming      33.085106   31.0    94  16.650525  51.936170   47.0    94   
Trail Run          38.088889   39.0    90  18.782399  50.988889   52.0    90   
Treadmill          37.826531   41.0    98  19.452308  45.877551   43.0    98   
Trekking           36.021277   36.0    94  17.754587  53.329787   51.5    94   
Walking            34.785714   33.0    98  17.722518  45.112245   44.0    98   

                            
                       std  
workout_type                
Cricket          27.513217  
Freestyle        29.580510  
Indoor Cycling   27.444582  
Open Water       29.897452  
Outdoor Cycling  28.542327  
Outdoor Running  28.641952  
Pool Swimming    26.498046  
Trail Run        30.491293  
Treadmill        26.322060  
Trekking         27.013891  
Walking          27.048975  

[11 rows x 60 columns]